adafruit_display_shapes/triangle.py

```diff
@@ -0,0 +1,218 @@
+"""
+`triangle`
+================================================================================
+
+Various common shapes for use with displayio - Triangle shape!
+
+
+* Author(s): Melissa LeBlanc-Williams
+
+Implementation Notes
+--------------------
+
+**Software and Dependencies:**
+
+* Adafruit CircuitPython firmware for the supported boards:
+  https://github.com/adafruit/circuitpython/releases
+
+"""
+
+import displayio
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_Display_Shapes.git"
+
+
+class Triangle(displayio.TileGrid):
+    # pylint: disable=too-many-arguments,invalid-name
+    """A triangle.
+
+    :param x0: The x-position of the first vertex.
+    :param y0: The y-position of the first vertex.
+    :param x1: The x-position of the second vertex.
+    :param y1: The y-position of the second vertex.
+    :param x2: The x-position of the third vertex.
+    :param y2: The y-position of the third vertex.
+    :param fill: The color to fill the triangle. Can be a hex value for a color or
+                 ``None`` for transparent.
+    :param outline: The outline of the triangle. Can be a hex value for a color or
+                    ``None`` for no outline.
+    """
+    def __init__(self, x0, y0, x1, y1, x2, y2, *, fill=None, outline=None):
+        # Sort coordinates by Y order (y2 >= y1 >= y0)
+        if y0 > y1:
+            y0, y1 = y1, y0
+            x0, x1 = x1, x0
+
+        if y1 > y2:
+            y1, y2 = y2, y1
+            x1, x2 = x2, x1
+
+        if y0 > y1:
+            y0, y1 = y1, y0
+            x0, x1 = x1, x0
+
+        smallest_x = x0
+        largest_x = x0
+
+        # Find the largest and smallest X values to figure out width for bitmap
+        if x1 > largest_x:
+            largest_x = x1
+
+        if x1 < smallest_x:
+            smallest_x = x1
+
+        if x2 > largest_x:
+            largest_x = x2
+
+        if x2 < smallest_x:
+            smallest_x = x2
+
+        height = y2 - y0 + 1
+        width = largest_x - smallest_x + 1
+
+        self._palette = displayio.Palette(3)
+        self._palette.make_transparent(0)
+        self._bitmap = displayio.Bitmap(width, height, 3)
+
+        if fill is not None:
+            self._helper(x0 - smallest_x, 0, x1 - smallest_x, y1 - y0, x2 - smallest_x, y2 - y0)
+            self._palette[2] = fill
+        else:
+            self._palette.make_transparent(2)
+
+        if outline is not None:
+            print("outline")
+            self._palette[1] = outline
+            self._line(x0 - smallest_x, 0, x1 - smallest_x, y1 - y0, 1)
+            self._line(x1 - smallest_x, y1 - y0, x2 - smallest_x, y2 - y0, 1)
+            self._line(x2 - smallest_x, y2 - y0, x0 - smallest_x, 0, 1)
+
+        super().__init__(self._bitmap, pixel_shader=self._palette, x=smallest_x, y=y0)
+
+    # pylint: disable=invalid-name, too-many-locals, too-many-branches
+    def _helper(self, x0, y0, x1, y1, x2, y2):
```

Comment:
What does this do?

Reply:
It draws the filled triangle. I can rename.